yt/fields/field_info_container.py

```diff
@@ -45,21 +54,31 @@ class FieldInfoContainer(UserDict):
     known_particle_fields: KnownFieldsT = ()
     extra_union_fields: Tuple[Tuple[str, str], ...] = ()
 
-    def __init__(self, ds, field_list, slice_info=None):
+    def __init__(self, ds, field_list: List[FieldKey], slice_info=None):
         super().__init__()
-        self._show_field_errors = []
+        self._show_field_errors: List[Exception] = []
         self.ds = ds
         # Now we start setting things up.
         self.field_list = field_list
         self.slice_info = slice_info
-        self.field_aliases = {}
-        self.species_names = []
-        if ds is not None and is_curvilinear(ds.geometry):
-            self.curvilinear = True
-        else:
-            self.curvilinear = False
+        self.field_aliases: Dict[FieldKey, FieldKey] = {}
+        self.species_names: List[FieldName] = []
         self.setup_fluid_aliases()
 
+    @property
+    def curvilinear(self) -> bool:
+        issue_deprecation_warning(
+            "FieldInfoContainer.curvilinear attribute is deprecated. "
+            "Please compare the internal dataset geometry directly to known Geometry enum members instead. ",
+            since="4.2",
+        )
+        geometry = self.ds.geometry
+        return (
```

Comment:
Sadly we can't (yet ?) because mypy doesn't understand that it's equivalent. I don't know if there's a good reason why, or if it can be added in the future, but as of now this is the only syntax that allows verifying exhaustiveness.

Reply:
but why the depracation? Having a shorthand function or property seems useful still.

Reply:
same answer as to why I'm proposing to deprecate `is_curvilinear`

Reply:
just going to reply here and resolve the other conversation :) 

I agree that deprecating the public `is_curvilinear` function makes some sense (for the reasons you gave in the other discussion). But it does seem nice to have some short-hand functions to consolidate some of these longer `geometry is .. or geometry is ...` blocks that get repeated.

Reply:
The purpose of these checks is to ba validated at type check time. Now, it may be possible to encapsulate part of them using e.g. [`typing.TypeGuard`](https://docs.python.org/3/library/typing.html#typing.TypeGuard), but:
1) it's more likely to _increase_ overall complexity than reducing it.
2) there is no indication that doing fancy combination like `StrEnum` + `TypeGuard` is already supported by mypy (`StrEnum` is brand new, most likely mypy will get smarter about it with time)
3) there is actually a way to simplify repeated `is Geometry...` checks, but it involves Python 3.10+ syntax:

```python
if geometry is Geometry.CARTESIAN or geometry is Geometry.SPECTRAL_CUBE:
    print("case 1")
elif (
    geometry is Geometry.POLAR
    or geometry is Geometry.CYLINDRICAL
    or geometry is Geometry.SPHERICAL
    or geometry is Geometry.GEOGRAPHIC
    or geometry is Geometry.GEOGRAPHIC_INTERNAL
):
    print("case 2")
else:
    assert_never(geometry)
```
is equivalent to
```python
match geometry:
    case Geometry.CARTESIAN | Geometry.SPECTRAL_CUBE:
        print("case 1")
    case Geometry.POLAR | Geometry.CYLINDRICAL | Geometry.SPHERICAL | Geometry.GEOGRAPHIC | Geometry.GEOGRAPHIC_INTERNAL:
        print("case 2")
    case _:
        assert_never(geometry)
```

Reply:
I guess what I mean to say is that I don't think we should try to outsmart mypy, and give it a chance to catch up before we attempt to refine the code. I'd rather have complex and verified code than a nice looking but unverifiable interface.
Right now, I don't think we can have both.

Reply:
@chrishavlin can we resolve this thread ?